new-log-viewer: Integrate latest clp-ffi-js with support for log-level filtering; Add log-level selector to `StatusBar`.

junhaoliao opened the pull request:
# References
<!-- Any issues or pull requests relevant to this pull request -->
new-log-viewer series: #45 #46 #48 #51 #52 #53 #54 #55 #56 #59 #60 #61 #62 #63 #66 #67 #68 #69 #70 #71 #72 #73 #74 #76

# Description
<!-- Describe what this request will change/fix and provide any details 
necessary for reviewers -->
1. Add log level selector component to StatusBar. Note no actual log level filtering is resulted from the selector changes, which is to be submitted in another PR.

# Validation performed
<!-- What tests and validation you performed on the change -->
1. Referred to $46, launched debug server and loaded the log-viewer in a browser.
2. Observed a select box with label `Log Level ^` in the Statusbar located in the bottom of the viewport. 
3. Clicked on the select box to open the select dropdown, and observed none of the options were selected by default.
4. Clicked on option "DEBUG" and observed all options higher than the "DEBUG" level (i.e., "INFO", "WARN", "ERROR", "FATAL") were selected. Also, the "Log Level" label on the select box was replaced with smaller icons which indicate the log levels that have been selected.
5. Hovered onto the icon "D" and observed a tooltip of "D" showing up. Clicked on the icon and observed the "D" disappeared, which could be also confirmed by opening the select dropdown again.
6. Opened the select dropdown and selected the checkbox next to "ERROR". Observed the "ERROR" option getting deselected and the other selected options remained selected. ~~Observed all options higher than the "ERROR" level were selected.~~
7. Opened the select dropdown again and selected the checkbox next to "TRACE". Observed the "TRACE" option getting selected in addition to the other selected options. ~~Observed all options higher than the "TRACE" level were selected.~~
8. Opened the select dropdown again and unchecked all selected options. Observed the icons in the select box was replaced by the "Log Level ^" label.
9. Opened the select dropdown and selected "NONE". Observed all options were selected. Clicked the "X" button on the select box and observed all options were cleared.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `LogLevelSelect` component for selecting multiple log levels with a user-friendly dropdown interface.
	- Added constants for log level management, including `INVALID_LOG_LEVEL_VALUE`, `LOG_LEVEL_NAMES`, `LOG_LEVEL_VALUES`, and `MAX_LOG_LEVEL`.

- **Bug Fixes**
	- Enhanced functionality for log level filtering and event mapping.

- **Chores**
	- Updated the version of the `clp-ffi-js` dependency for improved library features.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

junhaoliao committed:
Add log level selector component to StatusBar.

junhaoliao committed:
Add "Log Level" label as a Chip to LogLevelSelect component.

junhaoliao commented on new-log-viewer/src/utils/data.ts:
```diff
@@ -38,7 +38,45 @@ const getMapValueWithNearestLessThanOrEqualKey = <T>(
 };
 
+/**
+ * Creates an array of numbers in a specified range [startNum, startNum + steps].
```
This utility function was created with a reference to Python's build-in https://docs.python.org/3.12/library/stdtypes.html#range method.

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.css:
```diff
@@ -0,0 +1,8 @@
+    gap: 2px;
+}
+
+.log-level-select-render-value-box-label {
+    background-color: initial !important;
+}
```
The `!important` usage is to override the styles from JoyUI and therefore cannot be easily removed.

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/LogLevelChip.css:
```diff
@@ -0,0 +1,7 @@
+    --Chip-radius: 0;
+
+    font-family: "Roboto Mono", monospace !important;
+    font-weight: 600 !important;
```
ditto. See the discussion at https://github.com/y-scope/yscope-log-viewer/pull/77#discussion_r1776577098

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/LogLevelChip.css:
```diff
@@ -0,0 +1,7 @@
+    /* stylelint-disable-next-line custom-property-pattern */
+    --Chip-radius: 0;
+
+    font-family: "Roboto Mono", monospace !important;
```
ditto. See the discussion at https://github.com/y-scope/yscope-log-viewer/pull/77#discussion_r1776577098

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/LogLevelChip.tsx:
```diff
@@ -0,0 +1,66 @@
+    [LOG_LEVEL.FATAL]: "danger",
+});
+
+interface LogLevelChipProps {
+    name: string,
+    value: LOG_LEVEL,
+
+    onSelectedLogLevelsChange: (func: (value: LOG_LEVEL[]) => LOG_LEVEL[]) => void,
+}
```
sg. Renamed as `setter`.

junhaoliao committed:
Rename parameter in LogLevelChip's onSelectedLogLevelsChange callback.

junhaoliao committed:
Replace inline style with a CSS class for the dummy MenuItem.

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,109 @@
+            onChange={handleSelectChange}
+        >
+            {/* Add a dummy MenuItem to avoid the first Option receiving focus. */}
+            <MenuItem sx={{display: "none"}}/>
```
This is a JoyUI specific bug and the MenuItem was added as a mitigation.

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,109 @@
+            {/* Add a dummy MenuItem to avoid the first Option receiving focus. */}
+            <MenuItem sx={{display: "none"}}/>
+            {LOG_LEVEL_NAMES.map((logLevelName, index) => (
+                <Option
+                    key={logLevelName}
+                    value={index}
+                >
+                    {logLevelName}
+                </Option>
+            ))}
```
`Object.values(LOG_LEVEL)` returns not only the values in enum `LOG_LEVEL` but also the key names, which means we will need to perform extra filtering here.

Also, `LOG_LEVEL` is a numerical enum with values starting from 0 and therefore `index` is safe to be used.

junhaoliao committed:
Add comparator to .sort() - Apply suggestions from code review

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

junhaoliao committed:
Change select box behaviour to always enable the selected log level and above values on clicks.

junhaoliao committed:
Do not specify LogLevelChip tooltip color based on the log level; add "Click to deselect" prompt to the tooltip.

junhaoliao committed:
Limit LogLevelSelect listbox width.

junhaoliao commented on new-log-viewer/tsconfig.json:
```diff
@@ -17,11 +17,11 @@
     /* Specify a set of bundled library declaration files that describe the target runtime environment. */
     "lib": [
-      "ES2022",
+      "ES2023",
```
To use `Array.toReversed()`.

kirkrodrigues commented:
Although this is nicer, I think the behaviour still needs some refinement to be more intuitive.

# General refinement

1. **Differentiate UNKNOWN and ALL:** Currently, we're treating `NONE` as `ALL`, but I think it's more intuitive to rename `NONE` to `UNKNOWN` and treat it as the selector for log events where we couldn't parse a log level. Then we should also add an `ALL` selector to clear the filter.
2. **Don't allow deselecting individual levels from the chips**: The chips are neat but since clicking a chip causes it to disappear, it has asymmetric behaviour---you can click it to deselect the level but you can't click it again to re-select the level. I think it's better that the chips are simply used for display and the select/deselect logic is only done through the selector.
3. **Add a checkbox beside every selection option:** This would allow the user to choose between selecting individual levels using the checkboxes or clicking the label to select the range of log levels with >= importance of the desired level. If this is difficult, we could omit it until after the next CLP release.

junhaoliao committed:
Remove log-level deselection handling on the chips.

junhaoliao committed:
Rename log level `NONE` -> `UNKNOWN`; Add an "ALL" option to clear filters.

junhaoliao committed:
Add checkbox to LogLevelSelect to allow individual selections.

junhaoliao committed:
Specify listbox placement to avoid the listbox moving as the select button resizes.

junhaoliao committed:
Set max-width directly to 0 to prevent auto-resizing with the Select button; add docs.

junhaoliao committed:
Fix an issue where clicking on some Option edge deselects all options.

junhaoliao committed:
Extract `PlacementLeftTooltip` and `PlacementRightTooltip` components to handle tooltips with left and right placements respectively.

junhaoliao committed:
Refactor Tooltip components to use destructuring.

junhaoliao committed:
Extract components from LogLevelSelect.

junhaoliao committed:
Rename `LOG_LEVEL.NONE` -> `LOG_LEVEL.UNKNOWN`.

kirkrodrigues commented:
Very cool. Can we change a few more things (changed my mind on a few things after trying it)?

* Tooltips:
  * Checkbox: I think `+/- LEVEL` is more intuitive?
  * Option: The ^ looks sleek but I think there's potential for misinterpretation. How about "LEVEL and above"?
    * On that note, I think from a noisiness perspective, people usually think of "reducing the log level". So perhaps ordering the levels in reverse is more intuitive?
* Can we rename "ALL" to "Clear filters"? That feels clearer with the way it's currently designed.
  * In that case, we probably don't need the tool tip.

kirkrodrigues commented:
> So perhaps ordering the levels in reverse is more intuitive?
Obviously if we go with this, we should change the tooltip to "LEVEL and below"

junhaoliao commented:
>   * Checkbox: I think `+/- LEVEL` is more intuitive?

Right, it was to align the icons since different log levels have different number of characters in their names. On the other hand, the tooltips are not displayed all at the same time so it's less obvious if the icons have shifted a bit.

>   * Option: The ^ looks sleek but I think there's potential for misinterpretation. How about "LEVEL and above"?

"LEVEL and xxx" could be too long to be displayed on the right of the target element and cause the checkbox to be covered. Let me double check if there's a better way to handle this.

>     
>     * On that note, I think from a noisiness perspective, people usually think of "reducing the log level". So perhaps ordering the levels in reverse is more intuitive?

Sure, let's change this to "LEVEL and below" then.

> * Can we rename "ALL" to "Clear filters"? That feels clearer with the way it's currently designed.
>   
>   * In that case, we probably don't need the tool tip.

sg

junhaoliao committed:
In the checkbox tooltip, place +/- icon on the left of the LEVEL name.

kirkrodrigues commented:
Sorry, one more thing: One of my pet peeves is UI elements that move when things load. Right now, the log level filter is to the left of the log event number which means that when the log event number changes significantly, the log level filter will move. Can we put the log level filter on the right instead (like we had it in the old log viewer)?

We'll probably need to recolour the log event number to be something less emphasized. Why did we emphasize it anyway?

junhaoliao committed:
Rename "ALL" option to "Clear filters"; move "Clear filters" to the top of the list and reorder the log level options from "UNKNOWN" to "FATAL" from the top; remove the MenuItem mitigation.

junhaoliao committed:
Revert tsconfig changes now that we don't need to use `Array.toReversed()`.

junhaoliao committed:
Move LogLevelSelect to the right of the LogEventNum button in the status bar; reduce emphasis on the button.

junhaoliao committed:
Remove unused React import from LogLevelChip.tsx.

junhaoliao commented on new-log-viewer/src/typings/logs.ts:
```diff
@@ -8,9 +8,28 @@ enum LOG_LEVEL {
+);
+
+const MAX_LOG_LEVEL = Math.max(...LOG_LEVEL_VALUES);
+
+
```
```suggestion

```

junhaoliao commented on new-log-viewer/src/components/StatusBar/index.css:
```diff
@@ -3,6 +3,7 @@
     bottom: 0;
 
     display: flex;
+    flex-wrap: wrap;
```
This is to avoid the individual components in the status bar squeezing themselves when the browser window is too narrow.

One may not like that the components are wrapped outside of the viewport so that a scroll-down becomes necessary to access the wrapped components. If so, an issue can be submitted to keep track of the behaviour and we can look for a better solution in the future.

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,272 @@
+ */
+const ClearFiltersOption = ({onClick}: ClearFiltersOptionProps) => (
+    <Option
+        value={-1}
```
Shall we create something like `const INVALID_LOG_LEVEL_SELECT_OPTION_VALUE = -1;`?

* Or instead of using -1, would be make sense to use `LOG_LEVEL.UNKNOWN - 1`?
* I haven't tried but `null` might also work here.

junhaoliao commented on new-log-viewer/public/index.html:
```diff
@@ -10,6 +10,7 @@
     <link rel="preconnect" href="https://fonts.googleapis.com">
     <link rel="preconnect" crossorigin href="https://fonts.gstatic.com">
     <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Roboto&display=swap">
+    <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Roboto+Mono&display=swap">
```
I recently learned that we may want to limit the range of font weight imported to reduce the asset sizes. What do you think?

```suggestion
    <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Roboto+Mono:wght@300..700&display=swap">
```

Also shall we use `@import` for font imports? e.g.
```
<style>
@import url('https://fonts.googleapis.com/css2?family=Roboto+Mono:wght@300..700&display=swap');
</style>
```

junhaoliao replied:
1. I can't find one. With that said, it might be better to use a CDN for font loading rather than serving the font files directly from the log-viewer, so that loading of the log viewer can be sped up if the font has been loaded locally by another site. (It can be another story when we release the Log Viewer with Desktop UI frameworks like Electron. )
2. [Online critics](https://stackoverflow.com/questions/12316501/including-google-fonts-link-or-import) have said that `@import` defer loading of other resources, so we might not want to use `@import` given we have been managing all font loading in the same HTML file.

kirkrodrigues replied:
1. Gotcha. I guess we can specify a weight then, when we need to add fonts.
2. Gotcha.

junhaoliao committed:
Mitigate a rounding issue in JoyUI when calculating the listbox position.

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,273 @@
+ */
+const ClearFiltersOption = ({onClick}: ClearFiltersOptionProps) => (
+    <Option
+        value={-1}
```
I think the current behaviour for this option's code is a bit unintuitive. For the normal select options, we use data-value to determine the start of the range. But for this one, data-value is unset, so the start of the range ends up being NaN. In that case, perhaps we should just check whether data-value is set and if not, we clear the range? And in that case, it doesn't really matter what we set for value here.

junhaoliao replied:
sg

junhaoliao replied:
In fact, now I recall why I supplied this `-1` value - it's a mandatory attribute of `<Option/>`. Since we need the `<Option/>` here only because we want a button that appears the same as the other options, the value is unimportant here.

`null` works. Shall we keep it as `null`?

kirkrodrigues replied:
Sg

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.css:
```diff
@@ -0,0 +1,20 @@
+.log-level-select {
+    /* JoyUI has a rounding issue when calculating the listbox position, causing it to misjudge if
+      the  listbox will fit on the right side. To mitigate this, shift the select box 1px to the
```
```suggestion
      the listbox will fit on the right side. To mitigate this, we shift the select box 1px to the
```

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,273 @@
+ * @param props.rest
+ * @return
+ */
+const PlacementLeftTooltip = ({children, ...rest}: TooltipProps) => (
```
It seems like these PlacementXxxTooltips only save one line?

junhaoliao replied:
Right... Previously it was saving three lines. Let me inline it.

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,273 @@
+
+
+/**
+ * Renders a `<Tooltip/>` with its placement set to the left of the children element.
```
```suggestion
 * Renders a `<Tooltip/>` with its placement set to the left of its children.
```

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,273 @@
+);
+
+/**
+ * Renders a `<Tooltip/>` with its placement set to the right of the children element.
```
```suggestion
 * Renders a `<Tooltip/>` with its placement set to the right of its children.
```

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,273 @@
+}
+
+/**
+ * Renders an <Option/> in the <LogLevelSelect/> for selecting some log level and the level above
+ * it.
```
```suggestion
 * Renders an <Option/> in the <LogLevelSelect/> for selecting some log level and/or the levels
 * above it.
```

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,273 @@
+    logLevelName,
+    logLevelValue,
+    onCheckboxClick,
+}:LogSelectOptionProps) => {
```
```suggestion
}: LogSelectOptionProps) => {
```

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,272 @@
+                        onClick={onCheckboxClick}/>
+                </PlacementLeftTooltip>
+            </ListItemDecorator>
+            <PlacementRightTooltip
```
Right. It should it be placement left, right (if we want to be explicit)?

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,273 @@
+                    </Stack>
+                }
+            >
+                <ListItemContent data-value={logLevelValue}>
```
For my own understanding, could you explain why we need data-value on both the Option and this?

junhaoliao replied:
When `onCheckboxClick(ev)` is invoked, the event target can either be the Option (an `<ul/>`) or the ListItemContent (a `<div/>` as a child of the <ul/>`) depending on whether the user is clicking the edges of the option.

Having the data attribute on both elements allows the handler to read the value without too many changes in the handler itself.

kirkrodrigues replied:
Hm, I guess this is because we can't use currentTarget, right? If so, do you think it's worth adding a comment?

junhaoliao replied:
> this is because we can't use currentTarget

With the handler attached to the `<Select/>`, no. However, there is no real benefit of passing the handler to the `onChange` prop of `<Select/>` now we are not even utilizing the second parameter (newValue) of `onChange`. 

With the handler attached to the `<Option/>` directly, yes. I'll make the change.

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,273 @@
+                    <IconButton
+                        variant={"plain"}
+                        onClick={handleSelectClearButtonClick}
+                        onMouseDown={handleSelectClearButtonMouseDown}
```
What's this for?

junhaoliao replied:
To my recollection it attempted to avoid toggling the listbox when the button is clicked although I can no longer reproduce the behaviour with this handler removed now. 
Let me remove it altogether to avoid confusions.

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/index.tsx:
```diff
@@ -38,15 +39,17 @@ const StatusBar = () => {
             >
-                Log Event
-                {" "}
+                {"Log Event: "}
```
Do we need the colon? It makes it inconsistent with "Log level" where we can't add the colon.

kirkrodrigues commented on new-log-viewer/src/utils/data.ts:
```diff
@@ -38,7 +38,45 @@ const getMapValueWithNearestLessThanOrEqualKey = <T>(
         map.get(lowerBoundKey) as T;
 };
 
+/**
+ * Creates an array of numbers in a specified range [startNum, startNum + steps].
+ *
+ * @param start The value of the start parameter (or `0` if the parameter was not supplied).
+ * @param stop The value of the stop parameter.
+ * @param step The value of the step parameter (or `1` if the parameter was not supplied).
+ * - For a positive step, the contents of a range r are determined by the formula `r[i] = start +
+ * step*i` where `i >= 0` and `r[i] < stop`.
+ * - For a negative step, the contents of the range are still determined by the formula `r[i] =
+ * start + step*i`, but the constraints are `i >= 0` and `r[i] > stop`.
+ * @return An array of numbers from `start` to `stop` (exclusive) with a step of `step`.
+ * @throws {Error} if `step` is 0.
+ */
```
```suggestion
/**
 * Creates an array of numbers in the range `[start, stop)` with the `i`th element computed as
 * `range[i - 1] + step`.
 *
 * @param start
 * @param stop
 * @param step
 * @return The computed range.
 * @throws {Error} if `step` is 0.
 */
```

I think this relies a bit more on user intuition but is simpler to understand imo. Let me know if it's missing some detail we need.

kirkrodrigues commented on new-log-viewer/src/utils/data.ts:
```diff
@@ -38,7 +38,45 @@ const getMapValueWithNearestLessThanOrEqualKey = <T>(
+ * @return An array of numbers from `start` to `stop` (exclusive) with a step of `step`.
+ * @throws {Error} if `step` is 0.
+ */
+const range = (start: number, stop: Nullable<number> = null, step: number = 1): number[] => {
```
Personally, I would use `begin` and `end` for consistency with all the other ranges we have, but it's up to you if you want to follow Python.

kirkrodrigues commented on new-log-viewer/src/utils/data.ts:
```diff
@@ -38,7 +38,45 @@ const getMapValueWithNearestLessThanOrEqualKey = <T>(
+ * @throws {Error} if `step` is 0.
+ */
+const range = (start: number, stop: Nullable<number> = null, step: number = 1): number[] => {
+    const result: number[] = [];
```
Can we move this closer to its first use?

kirkrodrigues commented on new-log-viewer/src/utils/data.ts:
```diff
@@ -38,7 +38,45 @@ const getMapValueWithNearestLessThanOrEqualKey = <T>(
+    }
+    if (null === stop) {
+        stop = start;
+        start = 0;
```
```suggestion
```

I don't think we need this, do we?

junhaoliao replied:
If the user only provides a single argument which is to indicate the ending index, it's assumed that the starting index is 0. I'm not sure if there's a better way to name the parameters though.

Let me document the behaviour first.

kirkrodrigues replied:
Oh, oh, I see. Could we maybe do something like (this is just the idea, not the exact syntax, lol) `range(args: number | {start: number, stop: number, step?: number})`?

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/LogLevelChip.css:
```diff
@@ -0,0 +1,7 @@
+    --Chip-radius: 0;
+
+    font-family: "Roboto Mono", monospace !important;
+    font-weight: 600 !important;
```
Sure. Let's see if this is now better
![image](https://github.com/user-attachments/assets/2bba5a4e-ecf1-42e6-9196-6a096a9109f4)

junhaoliao committed:
Remove Roboto Mono font and adjust LogLevelChip styles.

junhaoliao committed:
Docs - Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao committed:
Docs & Reformat - Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao committed:
Document one-argument behaviour in `range`.

junhaoliao committed:
Inline PlacementXXXTooltip components.

junhaoliao committed:
Remove extraneous onMouseDown event handler.

junhaoliao committed:
Change ClearFiltersOption value to null.

junhaoliao committed:
Remove colon from Log Event button label.

junhaoliao committed:
Move the declaration of the `result` array closer to its usage within the `range` function.

junhaoliao committed:
Refactor `range` function parameter names `start` -> ``begin`, `stop` -> `end`.

junhaoliao committed:
Refactor `range` function signature to better support varied argument formats.

kirkrodrigues commented on new-log-viewer/src/utils/data.ts:
```diff
@@ -38,7 +38,55 @@ const getMapValueWithNearestLessThanOrEqualKey = <T>(
+ * Creates an array of numbers in the range `[begin, end)` with the `i`th element computed as
+ * `range[i - 1] + step`.
+ *
+ * If only one argument is supplied, the argument is interpreted as `end`, and `begin` is set to
+ * `0`.
```
```suggestion
 * If `args` is a number, it is interpreted as `end`, and `begin` is set to `0`.
```

Feel free to disagree.

kirkrodrigues commented on new-log-viewer/src/utils/data.ts:
```diff
@@ -38,7 +38,55 @@ const getMapValueWithNearestLessThanOrEqualKey = <T>(
+const range = (
+    args: number | {begin: number, end: number} | {begin: number, end: number, step: number}
+): number[] => {
+    // If only one argument is supplied, interpret it as `end` with `begin` set to 0.
```
```suggestion
    // If `args` is a number, interpret it as `end` with `begin` set to 0.
```

Ditto.

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,235 @@
+ */
+const ClearFiltersOption = ({onClick}: ClearFiltersOptionProps) => (
+    <Option
+        value={null}
```
`null` causes a console error.

junhaoliao replied:
My bad... Let's use a const with value `LOG_LEVEL.UNKNOWN - 1` instead.

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,235 @@
+        }
+
+        const target = ev.target as HTMLElement;
+        const selectedValue = Number(target.dataset.value);
```
`selectedValue` will be `NaN` for the clear filters option path. Do you want to add the check to see if `target.dataset.value` exists here?

junhaoliao committed:
Attach the select change handler to the <Option/>s instead; rename the handler.

junhaoliao committed:
Add check for undefined data-value attribute in LogLevelSelect.

junhaoliao committed:
Change log level select clear option value to INVALID_LOG_LEVEL_VALUE.

junhaoliao committed:
Docs - Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/LogLevelChip.tsx:
```diff
@@ -0,0 +1,54 @@
+const LOG_LEVEL_COLOR_MAP: Record<LOG_LEVEL, DefaultColorPalette> = Object.freeze({
+    [LOG_LEVEL.UNKNOWN]: "neutral",
+    [LOG_LEVEL.TRACE]: "neutral",
+    [LOG_LEVEL.DEBUG]: "neutral",
```
```suggestion
    [LOG_LEVEL.DEBUG]: "success",
```
shall we?
<img width="306" alt="image" src="https://github.com/user-attachments/assets/7052452c-18b4-4537-bf37-b3a27139546d">

junhaoliao commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/LogLevelChip.tsx:
```diff
@@ -0,0 +1,54 @@
+ */
+const LOG_LEVEL_COLOR_MAP: Record<LOG_LEVEL, DefaultColorPalette> = Object.freeze({
+    [LOG_LEVEL.UNKNOWN]: "neutral",
+    [LOG_LEVEL.TRACE]: "neutral",
+    [LOG_LEVEL.DEBUG]: "neutral",
```
```suggestion
    [LOG_LEVEL.TRACE]: "success",
    [LOG_LEVEL.DEBUG]: "success",
```
or
<img width="306" alt="image" src="https://github.com/user-attachments/assets/6ed20fdc-8c3c-40c2-ba6e-7446833baba8">

kirkrodrigues replied:
Sg

junhaoliao committed:
Change "TRACE" & "DEBUG" chips' color to "success".

kirkrodrigues commented on new-log-viewer/src/components/StatusBar/LogLevelSelect/index.tsx:
```diff
@@ -0,0 +1,240 @@
+                    placement: "top-end",
+                },
+            }}
+
+            // onChange={handleOptionClick}
```
```suggestion
```

kirkrodrigues reviewed (changes requested):
For the PR title, how about:

new-log-viewer: Add log-level selector to `StatusBar`.

junhaoliao committed:
Remove commented out code.

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

kirkrodrigues reviewed (dismissed):
I guess we should wait for the other log level filtering PRs to be merged before this one.

junhaoliao committed:
Merge branch 'main' into log-level-select

# Conflicts:
#	new-log-viewer/src/services/decoders/JsonlDecoder.ts
#	new-log-viewer/src/typings/logs.ts
#	new-log-viewer/src/utils/data.ts

junhaoliao committed:
Change `LOG_LEVEL.NONE` to `LOG_LEVEL.UNKNOWN`.

junhaoliao committed:
Stop propagation from checkbox click to the option.

junhaoliao committed:
Merge remote-tracking branch 'junhao/log-level-select' into log-level-select

junhaoliao committed:
Remove JsonlDecoder.ts as it should have been done when merging from main.

junhaoliao commented:
> (1) I am not sure the ability to select/deselect an individual level is actually working. It looks like it just selects all the >= levels regardless.

The issue was the click event on the checkbox got propagated to the option behind it, so effectively the option was clicked after toggling some checkbox. Fixed in  03b40c0fd4cc08c48201f57d33d2179ebbbad160

> (2) This is more of an idea for a future PR. I think too much work for this PR. If possible, would be nice if the hover effect (to darken) would actually apply on all the >= level, and for the checkbox would just be == level. Not sure if this is possible in joyUI.

I agree that's a great idea and should clarify all confusion to new users. It might not be simply done with JoyUI. We can take a look in the future PR.

> UnreIated, just warning that I think some of the merges from main broke this branch, I had to checkout an earlier to commit to load a file.

Right, fixed in b76ebef3bea6b2b03718a1c2790fda7336c14c3e

kirkrodrigues commented on new-log-viewer/src/utils/data.ts:
```diff
@@ -133,10 +133,57 @@ const getMapValueWithNearestLessThanOrEqualKey = <T>(
 };
 
 
```
```suggestion
```

kirkrodrigues commented on new-log-viewer/src/typings/logs.ts:
```diff
@@ -22,13 +22,38 @@ interface LogEvent {
+);
+
+const MAX_LOG_LEVEL = Math.max(...LOG_LEVEL_VALUES);
+
```
```suggestion
```

kirkrodrigues commented on new-log-viewer/src/typings/logs.ts:
```diff
@@ -22,13 +22,38 @@ interface LogEvent {
+
+
 const INVALID_TIMESTAMP_VALUE = 0;
 
```
```suggestion


```

kirkrodrigues reviewed (dismissed):
A few nits otherwise lgtm.

junhaoliao committed:
Add / remove empty lines - Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

davemarco committed:
Merge branch 'main' into HEAD

davemarco committed:
tie button and add ir support

davemarco committed:
fix lint

davemarco committed:
fix seizure

kirkrodrigues reviewed (approved):
For the PR title, how about:

```
new-log-viewer: Integrate latest clp-ffi-js with support for log-level filtering; Add log-level selector to `StatusBar`.
